dartagnan/src/main/java/com/dat3m/dartagnan/verification/model/event/EventModelManager.java

```diff
@@ -0,0 +1,163 @@
+                        atomicBlock.add(executionModel.getEventModelById(j));
+                    }
+                    atomicBlocks.add(atomicBlock);
+                }
+                executionModel.addAtomicBlocks(t, atomicBlocks);
+            }
+        }
+    }
+
+    private EventModel extractEvent(Event e, ThreadModel tm, int id) {
+        EventModel em;
+        if (e.hasTag(Tag.MEMORY)) {
+            Object addressObj = checkNotNull(
+                manager.evaluateByModel(context.address((MemoryEvent) e))
+            );
+            final BigInteger address = new BigInteger(addressObj.toString());
+            executionModel.addAccessedAddress(address);
+            
+            String valueString = String.valueOf(
+                manager.evaluateByModel(context.value((MemoryCoreEvent) e))
+            );
+            final BigInteger value = switch(valueString) {
+                // NULL case can happen if the solver optimized away a variable.
+                // This should only happen if the value is irrelevant, so we will just pick 0.
+                case "false", "null" -> BigInteger.ZERO;
+                case "true" -> BigInteger.ONE;
+                default -> new BigInteger(valueString);
+            };
+
+            if (e.hasTag(Tag.READ)) {
+                em = new LoadModel(e, tm, id, address, value);
+                executionModel.addAddressRead(address, (LoadModel) em);
+            } else if (e.hasTag(Tag.WRITE)) {
+                em = new StoreModel(e, tm, id, address, value);
+                executionModel.addAddressWrite(address, (StoreModel) em);
+            } else {
+                // Should never happen.
+                logger.warn("Event {} has Tag MEMORY but no READ or WRITE", e);
+                em = new MemoryEventModel(e, tm, id, address, value);
+            }
+
+        } else if (e.hasTag(Tag.FENCE)) {
+            final String name = ((GenericVisibleEvent) e).getName();
+            em = new FenceModel(e, tm, id, name);
+        } else if (e instanceof Assert assrt) {
+            em = new AssertModel(assrt, tm, id);
+        } else if (e instanceof Local local) {
+            em = new LocalModel(local, tm, id);
+        } else if (e instanceof CondJump cj) {
+            em = new CondJumpModel(cj, tm, id);
+        } else {
+            // Should never happen.
+            logger.warn("Extracting the event {} that should not be extracted", e);
+            em = new DefaultEventModel(e, tm, id);
+        }
+
+        executionModel.addEvent(e, em);
+        return em;
+    }
+
+    private boolean toExtract(Event e) {
+        // We extract visible events, Locals and Asserts to show them in the witness,
+        // and extract also CondJumps for tracking internal dependencies.
+        return eventFilter.apply(e)
+               || e instanceof Local
+               || e instanceof Assert
+               || e instanceof CondJump;
+    }
+}
```

Comment:
We have two places where we are using the same information, here and `ThreadModel.getEventModelsToShow()`.

I think we need this method because otherwise we would have to add Models for all of our events. But then, do we really need the method in the ThreadModel?

Reply:
This method is not good for a simple reason: it uses a configurable `eventFilter` but also hardcodes 3 other types of events. Either it is fully configurable and those other events should be part of `eventFilter` or it is not and we extract everything.